eras/babbage/impl/src/Cardano/Ledger/Babbage/PParams.hs

```diff
@@ -0,0 +1,533 @@
+updateField 9 = field (\x up -> up {_a0 = SJust x}) From
+updateField 10 = field (\x up -> up {_rho = SJust x}) From
+updateField 11 = field (\x up -> up {_tau = SJust x}) From
+updateField 14 = field (\x up -> up {_protocolVersion = SJust x}) From
+updateField 16 = field (\x up -> up {_minPoolCost = SJust x}) From
+updateField 17 = field (\x up -> up {_coinsPerUTxOWord = SJust x}) From
+updateField 18 = field (\x up -> up {_costmdls = x}) (D $ SJust <$> decodeCostModelMap)
+updateField 19 = field (\x up -> up {_prices = SJust x}) From
+updateField 20 = field (\x up -> up {_maxTxExUnits = SJust x}) From
+updateField 21 = field (\x up -> up {_maxBlockExUnits = SJust x}) From
+updateField 22 = field (\x up -> up {_maxValSize = SJust x}) From
+updateField 23 = field (\x up -> up {_collateralPercentage = SJust x}) From
+updateField 24 = field (\x up -> up {_maxCollateralInputs = SJust x}) From
+updateField k = field (\_x up -> up) (Invalid k)
+
+instance (Era era) => FromCBOR (PParamsUpdate era) where
+  fromCBOR =
+    decode
+      (SparseKeyed "PParamsUpdate" emptyPParamsUpdate updateField [])
+
+-- =================================================================
+
+-- | Update operation for protocol parameters structure @PParams
+updatePParams :: PParams era -> PParamsUpdate era -> PParams era
+updatePParams pp ppup =
+  PParams
+    { _minfeeA = fromSMaybe (_minfeeA pp) (_minfeeA ppup),
+      _minfeeB = fromSMaybe (_minfeeB pp) (_minfeeB ppup),
+      _maxBBSize = fromSMaybe (_maxBBSize pp) (_maxBBSize ppup),
+      _maxTxSize = fromSMaybe (_maxTxSize pp) (_maxTxSize ppup),
+      _maxBHSize = fromSMaybe (_maxBHSize pp) (_maxBHSize ppup),
+      _keyDeposit = fromSMaybe (_keyDeposit pp) (_keyDeposit ppup),
+      _poolDeposit = fromSMaybe (_poolDeposit pp) (_poolDeposit ppup),
+      _eMax = fromSMaybe (_eMax pp) (_eMax ppup),
+      _nOpt = fromSMaybe (_nOpt pp) (_nOpt ppup),
+      _a0 = fromSMaybe (_a0 pp) (_a0 ppup),
+      _rho = fromSMaybe (_rho pp) (_rho ppup),
+      _tau = fromSMaybe (_tau pp) (_tau ppup),
+      _protocolVersion = fromSMaybe (_protocolVersion pp) (_protocolVersion ppup),
+      _minPoolCost = fromSMaybe (_minPoolCost pp) (_minPoolCost ppup),
+      _coinsPerUTxOWord = fromSMaybe (_coinsPerUTxOWord pp) (_coinsPerUTxOWord ppup),
+      _costmdls = fromSMaybe (_costmdls pp) (_costmdls ppup),
+      _prices = fromSMaybe (_prices pp) (_prices ppup),
+      _maxTxExUnits = fromSMaybe (_maxTxExUnits pp) (_maxTxExUnits ppup),
+      _maxBlockExUnits = fromSMaybe (_maxBlockExUnits pp) (_maxBlockExUnits ppup),
+      _maxValSize = fromSMaybe (_maxValSize pp) (_maxValSize ppup),
+      _collateralPercentage = fromSMaybe (_collateralPercentage pp) (_collateralPercentage ppup),
+      _maxCollateralInputs = fromSMaybe (_maxCollateralInputs pp) (_maxCollateralInputs ppup)
+    }
+
+-- ===================================================
+-- Figure 1: "Definitions Used in Protocol Parameters"
```

Comment:
we can probably remove this comment, we don't know what figure it'l be in the babbage spec yet